### DISCUSSION TITLE
About Compound hierarchy question

### DISCUSSION THREAD

**Discussion by 0xF6** [bepu/bepuphysics2#221] [Q&A] [answered]
Hi

I have a need for a tree hierarchy with the preservation of the local position relative to the parent

Example of a hierarchy:
```go
|PlanetA (KinematicBody)
 \ Unit1 (DynamicBody)
 \ Unit2 (DynamicBody)
 \ ...
|PlanetB (KinematicBody)
 \ Unit1 (DynamicBody)
 \ ...
```

Units with dynamic body must have a position relative to their parent

Iam looking at Compound\BigCompound and the question arose, but how to add\delete childs without rebuilding Compound?

**Comment by RossNordby** [comment, marked as answer]
A `Compound` is just a list of children with no acceleration structure, so adding to/removing from it amounts to just modifying the child buffer to be whatever you want.

A `BigCompound` is built for larger numbers of children, and so has a `Tree` acceleration structure that will also need to be updated.
``` cs
static void AddChildToCompound(ref BigCompound compound, CompoundChild child, BufferPool pool, Shapes shapes)
{
    pool.Resize(ref compound.Children, compound.Children.Length + 1, compound.Children.Length);
    compound.Children[^1] = child;
    shapes.UpdateBounds(child.LocalPose, child.ShapeIndex, out var bounds);
    compound.Tree.Add(bounds, pool);
}
```
``` cs
static void RemoveChildFromCompoundAt(ref BigCompound compound, int childIndex, BufferPool pool)
{
    var movedChildIndex = compound.Tree.RemoveAt(childIndex);
    if (movedChildIndex >= 0)
    {
        compound.Children[childIndex] = compound.Children[movedChildIndex];    
    }
    //Shrinking the buffer takes care of 'removing' the now-empty last slot.
    pool.Resize(ref compound.Children, compound.Children.Length - 1, compound.Children.Length - 1);
}

```
`BigCompound.Tree.Add` and `BigCompound.Tree.Remove` keep the `Tree` in a valid state, but if you make a lot of changes over time, the tree quality may degrade. That would lead to worse query and collision performance. If this happens, you may want to call `BigCompound.Tree.RefitAndRefine(pool, refinementIndex);` where `refinementIndex` is just an integer that is used by the refiner to choose different parts of the tree to prioritize. In other words, just increment it each time you refine or something along those lines. This probably doesn't need to be done every single time any change is made, but if the topology of the `Tree` has changed a lot, it's probably worth it.

It's worth noting that the innards are exposed specifically to let people mess with them in ways that are convenient for whatever they're doing- this isn't the only way to handle modifications. In particular, if you needed to make very large numbers of modifications, doing so one by one like this will very likely be slower than a batched process like https://github.com/bepu/bepuphysics2/blob/master/Demos/DemoMeshHelper.cs#L126.

I'll probably add an equivalent to the functions above to the `Compound`/`BigCompound`, now that I've gotten a question like this more than once.

That said, the name "DynamicBody" implies that you might be expecting dynamic behavior (in the physics simulation sense of the word) in the compound children. Children of a compound can't be independently dynamic, and while the compound shapes can support children being moved around, it's not what they were built for. Depending on your intent, there might be a more direct approach.

**Reply by 0xF6** [reply]
Thanks

**Reply by RossNordby** [reply]
>I'll probably add an equivalent to the functions above to the Compound/BigCompound, now that I've gotten a question like this more than once.

707d4a2278dde919d743da7b9c0c4074c9e8f0b1

Also fixed an oopsy in the earlier function, though it's now redundant.
